test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MvcEndpointDataSourceTests.cs

```diff
@@ -166,5 +190,150 @@ public void ChangeToken_MultipleChangeTokenProviders_ComposedResult()
             var compositeChangeToken = Assert.IsType<CompositeChangeToken>(changeToken);
             Assert.Equal(2, compositeChangeToken.ChangeTokens.Count);
         }
+
+        private MvcEndpointInfo CreateEndpointInfo(
+            string name,
+            string template,
+            RouteValueDictionary defaults = null,
+            IDictionary<string, object> constraints = null,
+            RouteValueDictionary dataTokens = null)
+        {
+            var constraintResolver = new DefaultInlineConstraintResolver(Options.Create<RouteOptions>(new RouteOptions()));
+            return new MvcEndpointInfo(name, template, defaults, constraints, dataTokens, constraintResolver);
+        }
+
+        private ActionDescriptor CreateActionDescriptor(string controller, string action, string area = null)
+        {
+            return new ActionDescriptor
+            {
+                RouteValues =
+                    {
+                        ["controller"] = controller,
+                        ["action"] = action,
+                        ["area"] = area
+                    },
+                DisplayName = string.Empty,
+            };
+        }
+
+        [Fact]
+        public void DefaultMvcRoute()
+        {
+            // Arrange
+            var mockDescriptorProvider = new Mock<IActionDescriptorCollectionProvider>();
+            mockDescriptorProvider.Setup(m => m.ActionDescriptors).Returns(new ActionDescriptorCollection(new List<ActionDescriptor>
+            {
+                CreateActionDescriptor("TestController", "TestAction")
+            }, 0));
+
+            var dataSource = CreateMvcEndpointDataSource(mockDescriptorProvider.Object);
+            dataSource.ConventionalEndpointInfos.Add(CreateEndpointInfo(string.Empty, "{controller}/{action}/{id?}"));
+
+            // Act
+            dataSource.InitializeEndpoints();
+
+            // Assert
+            var endpoint = Assert.Single(dataSource.Endpoints);
+            var matcherEndpoint = Assert.IsType<MatcherEndpoint>(endpoint);
+
+            Assert.Equal("TestController/TestAction/{id?}", matcherEndpoint.Template);
+        }
+
+        [Fact]
+        public void WithActionInlineDefault()
```

Comment:
I suppose you are planning to write more unit tests to test this logic of expanding templates. Usually I like to look at tests to understand how the logic/experience works in general.

Reply:
Many more